### PR TITLE
[web] Unify JS configuration. Make it available from initEngine.

### PR DESCRIPTION
**Pull request by ditman**
### Changes

This PR does most of the configuration work (including a stretch goal) described in [this document](https://docs.google.com/document/d/1rFM0qmYds1JkntTruBh8eO72xbT2KoP9JdbizucAAwQ/edit).

In particular, it **extends the currently existing plumbing to pass and store Flutter engine configuration from JS (through the `initializeEngine` method).**

The configuration system now behaves like this:

1. The already existing `JsFlutterConfiguration` JS-interop class is reused in the flutter.js `EngineInitializer` object so config passing from HTML -> Flutter stays the same across: the global variable, and initEngine call.
2. Users are warned that the old JS global variables (`window.flutterConfiguration` and `window.flutterWebRenderer`) are now deprecated, and that they should use `engineInitializer.initializeEngine(config)` instead.
3. An assertion is thrown in apps that use BOTH configuration methods, so users remember to clean-up the globals in their index.html (in release, it uses whatever comes from the `initializeEngine` call).

### Demo

This PR, not only adds a `hostElement` parameter (unused for now), but also all the properties defined in the `configuration.dart` file. For now:

* renderer
* canvasKitBaseUrl
* canvasKitMaximumSurfaces
* canvasKitForceCpuOnly
* hostElement
* debugShowSemanticsNodes

Here's [a demo of an app running this code](https://dit-tests.web.app), and configuring everything from its index.html.

Note that the demo has a warning, because it's using a deprecated configuration style:

Screen Shot 2022-11-02 at 6 33 11 PM

The demo _does fail_ with an assertion in development mode:

Screen Shot 2022-11-02 at 6 41 02 PM

#### TL;DR API:

```js
//...
onEntrypointLoaded: async function(engineInitializer) {
  let appRunner = await engineInitializer.initializeEngine({
    renderer: 'canvaskit',
    canvasKitBaseUrl: 'canvaskit/',
    canvasKitMaximumSurfaces: 64,
    targetElement: document.querySelector('#flutter_target'),
     // ... other config ...
  });
  appRunner.runApp();
}
```

### Issues

* First step of https://github.com/flutter/flutter/issues/32329
* Fixes https://github.com/flutter/flutter/issues/114700
* Fixes https://github.com/flutter/flutter/issues/114701

### Testing

* Manually deployed: https://dit-tests.web.app
* Added some unit tests:
    *  configuration object new features
    * initialization process accepts and stores configuration

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/configuration.dart [hunk 1]
```diff
@@ -152,10 +194,29 @@ class FlutterConfiguration {
...
   /// ```
-  bool get debugShowSemanticsNodes => _js?.debugShowSemanticsNodes ?? _defaultDebugShowSemanticsNodes;
+  bool get debugShowSemanticsNodes => getConfigValue<bool>(
+    (JsFlutterConfiguration c) => c.debugShowSemanticsNodes,
+    _defaultDebugShowSemanticsNodes);
   static const bool _defaultDebugShowSemanticsNodes = bool.fromEnvironment(
     'FLUTTER_WEB_DEBUG_SHOW_SEMANTICS',
   );
+
+  /// Returns the [targetElement] in which the Flutter Application is supposed
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/configuration.dart:204]: `targetElement` should be **`hostElement`** everywhere, as agreed upon in the design doc.

File: lib/web_ui/lib/src/engine/js_interop/js_loader.dart [hunk 2]
```diff
@@ -61,17 +42,12 @@ abstract class FlutterEngineInitializer{
...
-abstract class InitializeEngineFnParameters {
-}
-
 /// Typedef for the function that initializes the flutter engine.
-typedef InitializeEngineFn = Promise<FlutterAppRunner?> Function([InitializeEngineFnParameters?]);
+///
+/// [JsFlutterConfiguration] comes from `../configuration.dart`. It is the same
+/// object that can be used to configure flutter "inline", through the
+/// (to be deprecated) `window.flutterConfiguration` object.
+typedef InitializeEngineFn = Promise<FlutterAppRunner?> Function([JsFlutterConfiguration?]);
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/js_interop/js_loader.dart:50]: This _marries_ flutter.js and configuration.dart. ⚭

File: lib/web_ui/lib/src/engine/configuration.dart [hunk 3]
```diff
@@ -51,14 +52,49 @@ void debugSetConfiguration(FlutterConfiguration configuration) {
...
+  /// Overrides are always added at position [0] of the array, so the order in
+  /// which this method is called (first with flutterConfiguration), then with
+  /// the runtimeConfig, matters.
+  void addConfigurationOverrides(JsFlutterConfiguration? overrides) {
+    if (overrides != null) {
+      _configurations.insert(0, overrides);
+    }
+  }
+
+  T getConfigValue<T>(T? Function(JsFlutterConfiguration) extractor, T defaultValue) {
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/configuration.dart:80]: This could be `private` or at best `@visibleForTesting`

File: lib/web_ui/lib/src/engine/initialization.dart [hunk 4]
```diff
@@ -68,8 +69,11 @@ void debugEmulateHotRestart() {
 /// Fully initializes the engine, including services and UI.
 Future<void> initializeEngine({
   AssetManager? assetManager,
+  JsFlutterConfiguration? jsParams,
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/initialization.dart:72]: Should be: `runtimeConfiguration`

File: lib/web_ui/lib/src/engine/configuration.dart [hunk 5]
```diff
@@ -51,14 +52,49 @@ void debugSetConfiguration(FlutterConfiguration configuration) {
...
+    try {
```
Comment (eyebrowsoffire) [inline lib/web_ui/lib/src/engine/configuration.dart:81]: The try/catch here is unfortunate. Could you instead do something like:
```
return _configurations.map(extractor).firstWhere((T value) => value != null, () => defaultValue);
```

**Review by eyebrowsoffire — APPROVED** [review]
Looks good, with a small nitpick.

One suggestion, take it or leave it: It might be nice if `FlutterConfiguration` and `JSFlutterConfiguration` shared the same base class with the different configuration getters on it. That way, if a new field is added to the configuration, the analyzer will tell you that you need to add it to both classes. I'm not sure if this gets weird with the static interop stuff, so if it's too much of a pain feel free to ignore.

File: lib/web_ui/lib/src/engine/configuration.dart [hunk 6]
```diff
@@ -51,14 +52,49 @@ void debugSetConfiguration(FlutterConfiguration configuration) {
...
+      final JsFlutterConfiguration config = _configurations
+          .firstWhere(
+            (JsFlutterConfiguration config) => extractor(config) != null
+          );
+      return extractor(config)!; // Guaranteed not null, because two lines above.
+    } on StateError {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/configuration.dart:87]: Avoid using throwing exceptions for normal control flow scenarios. There's nothing wrong with runtime configs missing a value and needing to default to the `defaultValue`. Therefore no exceptions should be thrown. Consider using `orElse`, or just a plain loop.

See also: https://dzone.com/articles/exceptions-as-controlflow-in-java

File: lib/web_ui/lib/src/engine/configuration.dart (hunk 3 above)
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/configuration.dart:80]: We should assume that some config attributes are accessed frequently, such as `debugShowSemanticsNodes` and `canvasKitMaximuSurfaces`. Therefore, we should aim for this function to be very fast, or avoid calling it repeatedly, e.g. by memoizing the return value at the call side. We should aim to be close to the speed of field access. As it is, it relies on iteration, exceptions, closures, and repeated extraction of values, probably making it 10000x slower than field access.

But also, as I mention in the previous comment, perhaps it's better to not support the merging of multiple config objects, but use either `window.flutterConfiguration` (for legacy reasons) or the shiny new runtime config API.

File: lib/web_ui/lib/src/engine/configuration.dart (hunk 5 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/configuration.dart:81]: This is gone now. Ideally I should have used [this `firstWhereOrNull`](https://api.flutter.dev/flutter/package-collection_collection/IterableExtension/firstWhereOrNull.html) that does exactly what I needed (but I wasn't sure if the engine was already depending on the collection package)

File: lib/web_ui/lib/initialization.dart [hunk 7]
```diff
@@ -86,11 +86,11 @@ Future<void> webOnlyWarmupEngine({
   }
   if (autoStart) {
     // The user does not want control of the app, bootstrap immediately.
-    print('Flutter Web Bootstrap: Auto');
+    engine.domWindow.console.debug('Flutter Web Bootstrap: Auto.');
     await bootstrap.autoStart();
   } else {
     // Yield control of the bootstrap procedure to the user.
-    print('Flutter Web Bootstrap: Programmatic');
+    engine.domWindow.console.debug('Flutter Web Bootstrap: Programmatic.');
```
[the remark this answers is not in the document]
Comment (ditman) [reply lib/web_ui/lib/initialization.dart:93]: I thought I had an issue for this one, but couldn't find it... I guess it'll randomly show up in a couple of years :P

File: lib/web_ui/lib/src/engine/configuration.dart [hunk 8]
```diff
@@ -51,14 +52,49 @@ void debugSetConfiguration(FlutterConfiguration configuration) {
 class FlutterConfiguration {
   /// Constructs a configuration from a JavaScript object containing
   /// runtime-supplied properties.
-  FlutterConfiguration(this._js);
+  FlutterConfiguration(JsFlutterConfiguration? config) {
+    addConfigurationOverrides(config);
```
[the remark this answers is not in the document]
Comment (ditman) [reply lib/web_ui/lib/src/engine/configuration.dart:56]: I made this "crashy" if the user is both passing configuration through the global JS variable, and the initEngine API.

File: lib/web_ui/test/canvaskit/initialization/initialization_stores_config.dart [hunk 9]
```diff
@@ -0,0 +1,26 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (ditman) [reply lib/web_ui/test/canvaskit/initialization/initialization_stores_config.dart:1]: It totally does! I ran it via the command line with the full path, and didn't realize that I was missing the suffix. Good catch!

File: lib/web_ui/lib/src/engine/configuration.dart (hunk 6 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/configuration.dart:87]: This should have been a `firstWhereOrNull` or a plain loop, but now it's gone forever!

File: lib/web_ui/lib/src/engine/configuration.dart (hunk 3 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/configuration.dart:80]: You're right, I wasn't sure if any of these were used in a tight render loop. I've moved all the asserting/warning to the initialization code, so the retrieval is as fast as it can be (I think I actually reverted most of my changes wrt retrieving data, other than renaming the internal field).

File: lib/web_ui/lib/src/engine/app_bootstrap.dart [hunk 10]
```diff
@@ -4,19 +4,23 @@
 
 import 'package:js/js.dart';
 
+import 'configuration.dart';
 import 'js_interop/js_loader.dart';
 import 'js_interop/js_promise.dart';
 
+/// The type of a function that initializes an engine (in Dart)
```
[the remark this answers is not in the document]
Comment (ditman) [reply lib/web_ui/lib/src/engine/app_bootstrap.dart:11]: Please, do +1 here: https://github.com/dart-lang/linter/issues/1433 :P

**Comment by ditman** [comment]
> It might be nice if FlutterConfiguration and JSFlutterConfiguration shared the same base class with the different configuration getters on it. That way, if a new field is added to the configuration, the analyzer will tell you that you need to add it to both classes. I'm not sure if this gets weird with the static interop stuff, so if it's too much of a pain feel free to ignore.

@eyebrowsoffire I **think** that extending js-interop classes with dart classes (and vice versa?) is a big no-no. (Maybe that's why the FlutterConfiguration currently just wraps the JsFlutterConfiguration?)

This is a great suggestion, though! (I'd try it if I didn't have the suspicion that it can break wasm interop :P)

**Comment by ditman** [comment]
I think I addressed all the comments in the PR. PTAL @yjbanov, @eyebrowsoffire! And thanks for the review!

**Review by mdebbar — APPROVED** [review]
LGTM

File: lib/web_ui/lib/src/engine/configuration.dart [hunk 11]
```diff
@@ -5,10 +5,17 @@
...
+/// JS method, with an (optional) object of type [JsFlutterConfiguration].
 ///
-/// Example:
+/// This library also supports the legacy method of setting a plain JavaScript
+/// object set as the `flutterConfiguration` property of the top-level `window`
+/// object, but that approach is now deprecated and will warn users.
+///
+/// Both methods are **disallowed** to be used at the same time.
+///
+/// Example (Before):
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/configuration.dart:18]: I'd remove the "before" parts, let's boldly step into the new world! :)

File: lib/web_ui/lib/src/engine/configuration.dart [hunk 12]
```diff
@@ -18,47 +25,101 @@
...
-  FlutterConfiguration(this._js);
+  /// Constructs an unitialized configuration object.
+  @visibleForTesting
+  FlutterConfiguration();
 
-  final JsFlutterConfiguration? _js;
+  /// Constucts a "tainted by JS globals" configuration object.
+  ///
+  /// This is deprecated, and warns the user about the new API.
+  FlutterConfiguration.fromJsGlobals(JsFlutterConfiguration? config) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/configuration.dart:80]: I'd call this constructor `legacy`.

File: lib/web_ui/lib/src/engine/configuration.dart [hunk 13]
```diff
@@ -18,47 +25,101 @@
...
+    if (config != null) {
+      domWindow.console.warn('window.flutterConfiguration is now deprecated.\n'
+      'Use engineInitializer.initializeEngine(config) instead.\n'
+      'See: https://docs.flutter.dev/development/platform-integration/web/initialization');
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/configuration.dart:84]: Put the `console.warn` in an `assert` so we don't ship the long stings in release builds (applies to the one below too).

File: lib/web_ui/lib/src/engine/configuration.dart [hunk 14]
```diff
@@ -18,47 +25,101 @@
...
+  ///
+  /// This method is called by the engine initialization process, through the
+  /// [initEngineServices] method.
+  ///
+  /// This method throws an AssertionError, if the _configuration object has
+  /// been set to anything non-null through the [FlutterConfiguration.fromJsGlobals]
+  /// constructor.
+  void setUserConfiguration(JsFlutterConfiguration? configuration) {
+    if (configuration != null) {
+      assert(!_createdFromJsGlobals, 'Do not mix-and-match configuration styles. '
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/configuration.dart:108]: It's unclear what "mix-and-match" means. I'd say "Using both the legacy window.flutterConfiguration and initialization configuration simultaneously is not supported."

File: lib/web_ui/lib/src/engine/configuration.dart [hunk 15]
```diff
@@ -18,47 +25,101 @@
...
+        'Use engineInitializer.initializeEngine(config) instead of '
+        'window.flutterConfiguration.');
+      assert(_requestedRendererType == null || configuration.renderer == null,
+        'Do not mix-and-match configuration styles. '
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/configuration.dart:112]: ditto re "mix-and-match"

**Review by yjbanov — APPROVED** [review]

**Comment by yjbanov** [comment]
> It might be nice if FlutterConfiguration and JSFlutterConfiguration shared the same base class

@eyebrowsoffire @ditman I've always been suspicious about using inheritance for classes used to pass data, as opposed to delivering behavior. There are some expectations that come with inheritance that may not be appropriate. For example, there's the Liskov substitution principle which doesn't make sense for this case. I've always preferred composition for data over inheritance.

File: lib/web_ui/lib/src/engine/configuration.dart (hunk 11 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/configuration.dart:18]: I disagree, I'll remove these docs when we clean up the code. What I'm going to do, however, is move the new style to the top (and not call it "after"/"new" or anything), and move the legacy style underneath.

File: lib/web_ui/lib/src/engine/configuration.dart (hunk 14 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/configuration.dart:108]: Final text:

"Use engineInitializer.initializeEngine(config) only. Using the (deprecated) window.flutterConfiguration and initializeEngine configuration simultaneously is not supported."

**Comment by ditman** [comment]
Rebased to see if it heals the failing "impeller" tests.
